Trig dev

trigaut opened the pull request:
Adds :
- SDP
- A one storage example to test the features
- A jupyter notebook for a PHES management problem

TODO : 
- Switch from CamelCase to pep8
- Iterators for SDP
- Parallelisation for SDP
- Integrate barycentre for interpolation in SDP

Commit:
[DEV] Add basic two dams example

Commit:
[UPD] Update comments and minor fixes

Commit:
[DEV] Add SDP algorithm working only for compatible states, controls and uncertainties discretization. TODO: interpolation

Commit:
[FIX] Fixed various bugs in SDP

Commit:
[DEV] Add dams example using SDP (restrained to a small number of stages, small bounds and large discretization)

Commit:
[DEV] Add sdp algorithm for hazard-decision case and rename the decision-hazard case

Commit:
[UPD] factorize function that converts variables to index, Add a sampling function to simulate, Add HD case to twodams.jl

Commit:
[FIX] Fix some bugs in SDP

Commit:
[FIX] Add a one storage example to compare SDDP and SDP

Commit:
[UPD] Add small comments to SDPoptimize TODO : More comments

Commit:
[DEV] Reunion of SDPHD and SDPDH cases into one function, add nearest neighbor function to avoid exact discretization, change onedam.jl accordingly TODO delete some parameters in SDPparameters structure, make a onedam example that can be solved by SDPDH

Commit:
[DEV] Change value function structure, fix bugs in nearest neighbors, add a function to compute value functions barycentre

Commit:
[FIX] Solved silly forward in time error

Commit:
[DEV] Implemented a comparison between sdp HD and DH for the single dam example

Commit:
[DEV] Merging trig_dev and basic_dam

Commit:
[DEV] Add SDDP to onedam example

frapac committed:
[FIX] fix number of stages

Stages correspond to states, and include terminal state

Commit:
[DEV] Merging hotfix to trig_dev branch

Commit:
[DEV] Add SDDP-SDP comparison in the single dam example

Commit:
[UPD] Simplfied SDPparams constructor and add simple two dams example

Commit:
[DEV] Add PHES notebook and minors fixes in SDP

Commit:
[UPD] Change image location in PHES notebook

Commit:
[UPD] Changed staes and controls management in SDP for consistency

Commit:
[UPD] Add bounds to SDDP in onadam.jl example

frapac committed:
[UPD] Add ProgressMeter to REQUIRE

frapac committed:
Merge branch 'trig_dev' of https://github.com/leclere/StochDynamicProgramming.jl into trig_dev

frapac commented on examples/onedam.jl:
```diff
@@ -0,0 +1,366 @@
+srand(2713)
```
I think this module is a bit redundant with dam.jl

trigaut replied:
I should maybe add it to unit tests?

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+    return index
+end
+
+function variable_from_index( variable_ind::Int,
```
Could you comment this function?

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+Convert the state and control tuples (stored as arrays) of the problem into integers and vice versa
+
+"""
+function index_from_variable( variable::Array,
```
Could you detail the parameters and the output of this function?

frapac replied:
and we should think about using a decorator for these two functions (or use metaprogramming instead)

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+    j = 1;
+
+    for i = 1:length(variable)
+        index = index + j * (floor( Int, ( variable[i] - lower_bounds[i] ) / variable_steps[i] ))
```
`for v in variables` : use the modern way!

trigaut replied:
I did this because I needed to iterate over 3 other arrays as well. Should I prefer 4 iterators?

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+            variable[i] = lower_bounds[i] + variable_steps[i] * ((index % variable_sizes[i]))
+            index += -(index % variable_sizes[i])
+            index = index / variable_sizes[i]
+        end
```
index /= ...

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+Compute nearest neighbor of a continuous variable in the discrete variable space
+
+"""
+function nearest_neighbor( variable::Array,
```
Could you detail the parameters of this function?

frapac commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,466 @@
+        for i = 2:length(variable)
+            K=K*variable_sizes[i-1]
+            neighbors0 = copy(neighbors)
+            for j in neighbors0
```
This name is not so explicit

trigaut replied:
Which one? K?

frapac commented on src/objects.jl:
```diff
@@ -72,6 +72,29 @@ type PiecewiseLinearCostSPmodel <: SPModel
     end
 end
 
+type DPSPmodel <: SPModel
```
DPSP?

trigaut replied:
Dynamic Programming Stochastic Program?

leclere replied:
GenericSPmodel ? My point is : SDPoptimize should be able to solve the different SPmodel if we want to compare the algorithms.

Commit:
[UPD] Add comments to SDPoptimize

Commit:
[UPD] Update some not julia style lines

Commit:
[UPD] Updated the copyright

Commit:
[UPD] Changed the name of the PHES notebook

Commit:
[UPD] Change name of PHES notebook

trigaut committed:
[FIX] Fixed a reasoning error in the not yet used function barycentre

leclere commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,581 @@
+    #Define the loops order in sdp
+    if (param.infoStructure == "DH")
+
+        println("Starting stochastic dynamic programming
```
if display

leclere commented on src/SDPoptimize.jl:
```diff
@@ -0,0 +1,581 @@
+
+    else
+
+        println("Starting stochastic dynamic programming
```
if display

trigaut committed:
[FIX] Add if display in SDP

trigaut committed:
[FIX] Add forgotten end if

Commit:
[DEV] Add unit tests for SDP
	- Add a 1e-10 to index_from_variable to avoid floor errors
	- Delete export index_from_variable and variable_from_index

Commit:
[UPD] Add test barycentre for SDP

frapac committed:
[FIX] Correct number of stages in aleas generation

+ update size of JuMP.Model vector
+ fix examples

Commit:
Merge remote-tracking branch 'origin/hotfix-stage-number' into trig_dev

Commit:
[UPD] Change stages handling in PHES notebook

trigaut committed:
[DEV] Chage DPSPmodel name and add constructors to convert from linar spmodel

trigaut committed:
[FIX] Change tests to fix continuous integration issues

trigaut committed:
[FIX] Change 0. to 0 in onesteponealea to fix nighty build issue

trigaut committed:
[UPD] Add constructor to build StochDynProgModel from PiecewiseLinearCostSPmodel
